csp/impl/types/common_definitions.py

```diff
@@ -94,9 +122,34 @@ def __new__(cls, typ, shape: Optional[Union[List, int, str]] = None, shape_of: O
             # if shape is required, it will be enforced in the parser
             kwargs["shape"] = None
             kwargs["shape_func"] = None
+
         return type("OutputBasket", (OutputBasket,), kwargs)
 
 
+# Add core schema to OutputBasket
+def __get_pydantic_core_schema__(cls, source_type, handler):
+    from pydantic_core import core_schema
+
+    def validate_shape(v, info):
+        shape = cls.shape
+        # Allow the context to override the shape, for the cases where shape references an input variable name
+        # and so is not known until later
+        if info.context and hasattr(info.context, "shapes"):
+            override_shape = info.context.shapes.get(info.field_name)
+            if override_shape is not None:
+                shape = override_shape
+        if isinstance(shape, int) and len(v) != shape:
```

Comment:
Good catch, will fix.

Reply:
Actually, the fix here should be in the `__new__` function of `OutputBasket`, so that it raises if you try to construct a dict output basket with int shape or a list output basket with list or tuple shape (i.e. separate from any pydantic validation).